concent_api/core/payments/sci_callback.py

```diff
@@ -0,0 +1,260 @@
+        # Connection with Middleman times out before a complete response is received.
+        except socket.timeout as exception:
+            raise SCICallbackTimeoutError() from exception
+
+
+def deserialize_response_and_handle_errors(raw_response: bytes, request_id: int) -> SignedTransaction:
+    """
+    Deserialize received Frame and its payload and handle related errors.
+
+    Returns SignedTransaction message.
+    """
+
+    assert isinstance(raw_response, bytes)
+    assert isinstance(request_id, int)
+
+    try:
+        deserialized_message = GolemMessageFrame.deserialize(
+            raw_message=raw_response,
+            public_key=settings.CONCENT_PUBLIC_KEY,
+        )
+    # Received data frame is invalid or the signature does not match its content.
+    except MiddlemanProtocolError as exception:
+        raise SCICallbackFrameError() from exception
+    # If the received Golem message is malformed.
+    except MessageError as exception:
+        raise SCICallbackPayloadError() from exception
+
+    # Request ID in the received data frame does not match the one in the data frame that was sent.
+    if deserialized_message.request_id != request_id:
+        raise SCICallbackRequestIdError('MiddleMan response request_id does not match requested.')
+
+    # If the received frame is ErrorFrame.
+    if isinstance(deserialized_message, ErrorFrame):
+        raise SCICallbackPayloadError(
+            'Received frame is ErrorFrame.'
+        )
+
+    # If the received Golem message is TransactionRejected.
+    if isinstance(deserialized_message.payload, TransactionRejected):
+        raise SCICallbackPayloadError(
+            'Received frame is contains Golem message TransactionRejected.'
+        )
+
+    # If the received Golem message is anything else than SignedTransaction, TransactionRejected or and ERROR frame.
+    if not isinstance(deserialized_message.payload, SignedTransaction):
+        raise SCICallbackPayloadError('Received frame payload is not Golem message SignedTransaction instance.')
+
+    # If the received Golem message is not signed by the Signing Service.
+    try:
+        deserialized_message.payload.verify_signature(settings.SIGNING_SERVICE_PUBLIC_KEY)
+    except InvalidSignature as exception:
+        raise SCICallbackPayloadSignatureError() from exception
+
+    return deserialized_message.payload
+
+
+def verify_data_and_signature(signed_transaction: SignedTransaction, transaction: Transaction) -> None:
+    """
+    Verify received SignedTransaction data and signature and handle related errors.
+
+    1. Check if received SignedTransaction data match original Ethereum Transacion.
+    2. Check if received signature is created for given message with given keys pair.
+    """
+
+    assert isinstance(signed_transaction, SignedTransaction)
+    assert isinstance(transaction, Transaction)
+
+    # The transaction data in SignedTransaction does not match the information we sent.
```

Comment:
this comment seems unnecessary - the code is clear about itself ;)

Reply:
:+1: